### PR TITLE
Added matrix types that more closely resemble WGSL.

### PR DESCRIPTION
**Pull request by iwoplaza**

### CODE REVIEW

File: packages/typegpu/src/data/matrix.ts [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+  vec2i,
+  vec2u,
+  vec3f,
+  vec3i,
+  vec3u,
+  vec4f,
+  vec4i,
+  vec4u,
+  type vecBase,
+} from './vector';
```
Comment (reczkok) [inline packages/typegpu/src/data/matrix.ts]:
```suggestion
import { vec2f, vec3f, vec4f, type vecBase } from './vector';
```

File: packages/typegpu/src/data/matrix.ts [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+  makeColumn(x: number, y: number, z: number): vec3i {
+    return vec3i(x, y, z);
+  }
+}
+
+class mat3x3uImpl extends mat3x3Impl<vec3u> implements mat3x3u {
+  makeColumn(x: number, y: number, z: number): vec3u {
+    return vec3u(x, y, z);
+  }
+}
```
Comment (reczkok) [inline packages/typegpu/src/data/matrix.ts]:
```suggestion
```

File: packages/typegpu/src/data/matrix.ts [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+  makeColumn(x: number, y: number, z: number, w: number): vec4i {
+    return vec4i(x, y, z, w);
+  }
+}
+
+class mat4x4uImpl extends mat4x4Impl<vec4u> implements mat4x4u {
+  makeColumn(x: number, y: number, z: number, w: number): vec4u {
+    return vec4u(x, y, z, w);
+  }
+}
```
Comment (reczkok) [inline packages/typegpu/src/data/matrix.ts]:
```suggestion
```

File: packages/typegpu/src/data/matrix.ts [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+  makeColumn(e0: number, e1: number): vec2i {
+    return vec2i(e0, e1);
+  }
+}
+
+class mat2x2uImpl extends mat2x2Impl<vec2u> implements mat2x2u {
+  makeColumn(e0: number, e1: number): vec2u {
+    return vec2u(e0, e1);
+  }
+}
```
Comment (reczkok) [inline packages/typegpu/src/data/matrix.ts]:
```suggestion
```

File: packages/typegpu/src/data/matrix.ts [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+
+export interface mat2x2<TColumn> extends matBase<TColumn> {
+  [0]: TColumn;
+  [1]: TColumn;
+  [idx: number]: TColumn | undefined;
+}
+
+export interface mat2x2f extends mat2x2<vec2f> {}
+export interface mat2x2i extends mat2x2<vec2i> {}
+export interface mat2x2u extends mat2x2<vec2u> {}
```
Comment (reczkok) [inline packages/typegpu/src/data/matrix.ts]:
```suggestion
```

File: packages/typegpu/src/data/matrix.ts [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+export interface mat3x3<TColumn> extends matBase<TColumn> {
+  [0]: TColumn;
+  [1]: TColumn;
+  [2]: TColumn;
+  [idx: number]: TColumn | undefined;
+}
+
+export interface mat3x3f extends mat3x3<vec3f> {}
+export interface mat3x3i extends mat3x3<vec3i> {}
+export interface mat3x3u extends mat3x3<vec3u> {}
```
Comment (reczkok) [inline packages/typegpu/src/data/matrix.ts]:
```suggestion
```

File: packages/typegpu/src/data/matrix.ts [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+  [0]: TColumn;
+  [1]: TColumn;
+  [2]: TColumn;
+  [3]: TColumn;
+  [idx: number]: TColumn | undefined;
+}
+
+export interface mat4x4f extends mat4x4<vec4f> {}
+export interface mat4x4i extends mat4x4<vec4i> {}
+export interface mat4x4u extends mat4x4<vec4u> {}
```
Comment (reczkok) [inline packages/typegpu/src/data/matrix.ts]:
```suggestion
```

File: packages/typegpu/src/data/matrix.ts [hunk 8]
```diff
@@ -0,0 +1,423 @@
...
+  vec3f,
+  vec3i,
+  vec3u,
+  vec4f,
+  vec4i,
+  vec4u,
+  type vecBase,
+} from './vector';
+
+interface MatSchemaOptions<T, TVec extends vecBase> {
```
Comment (reczkok) [inline packages/typegpu/src/data/matrix.ts:20]: Should probably be restricted to float vectors.

File: packages/typegpu/tests/vector.test.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+import { vec2f, vec3f, vec4f } from '../src/data';
+
+describe('vec2f', () => {
+  it('should create a zero 2d vector', () => {
+    const zero = vec2f();
+    expect(zero.x).toEqual(0);
+    expect(zero.y).toEqual(0);
+  });
+
+  it('should create a 3d vector with the given elements', () => {
```
Comment (reczkok) [inline packages/typegpu/tests/vector.test.ts]:
```suggestion
  it('should create a 2d vector with the given elements', () => {
```

File: packages/typegpu/tests/vector.test.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    const zero = vec2f(1, 2);
+    expect(zero.x).toEqual(1);
+    expect(zero.y).toEqual(2);
+  });
+
+  it('should create a 3d vector from the given scalar element', () => {
```
Comment (reczkok) [inline packages/typegpu/tests/vector.test.ts]:
```suggestion
  it('should create a 2d vector from the given scalar element', () => {
```

File: packages/typegpu/tests/vector.test.ts [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    const zero = vec2f(5);
+    expect(zero.x).toEqual(5);
+    expect(zero.y).toEqual(5);
+  });
+
+  it('should encode a 3d vector', () => {
```
Comment (reczkok) [inline packages/typegpu/tests/vector.test.ts]:
```suggestion
  it('should encode a 2d vector', () => {
```

File: packages/typegpu/src/data/matrix.ts (hunk 8 above)
Comment (iwoplaza) [reply packages/typegpu/src/data/matrix.ts:20]: I do not have a base-type for float vectors. This is hidden to the end-user anyway, so until we accidentally create non-float schemas again, we should be okay.

**Pull request merged**